### DISCUSSION TITLE
Is there an function call to show the information about a connected device in host mode?

### DISCUSSION THREAD

**Discussion by micgresham** [hathach/tinyusb#1259] [General]
I am using the host/cdc_msc_hid example on a Pico RP2040 to connect an external USB-TTL adapter.  I also have the Pico configured to dump stdio out UART0.  

I would like to get detailed information when a device (USB-TTL) adapter is connect.  I am looking for something similar to the output of the linux lsusb -v command.

Any help is appreciated.

-michael

**Comment by hathach** [comment]
which specific info you want to get ?

**Comment by micgresham** [comment]
I want to see what is currently attached to the host controller (rp2040).  I will also be testing this with an STMf4xx controller and TinyUSB.
Mostly just looking for the basics:


  Product ID:                                    0x6001

  Vendor ID:                                     0x0403  (Future Technology Devices International Limited)

  Version:                                          6.00

  Serial Number:                              AB0KU7UZ

  Speed:                                           Up to 12 Mb/s

  Manufacturer:                                FTDI

I am working with a new CDC interface using the Silicon Labs CP2102 <https://www.silabs.com/documents/public/data-sheets/CP2102-9.pdf> and I am not sure it is coming up properly.  It is a USB-C attached serial console built into a piece of equipment.

Thanks,

-michael



From: Ha Thach ***@***.***>
Reply-To: hathach/tinyusb ***@***.***>
Date: Wednesday, December 22, 2021 at 12:52 AM
To: hathach/tinyusb ***@***.***>
Cc: micgresham ***@***.***>, Author ***@***.***>
Subject: Re: [hathach/tinyusb] Is there an function call to show the information about a connected device in host mode? (Discussion #1259)


which specific info you want to get ?

—
Reply to this email directly, view it on GitHub<https://github.com/hathach/tinyusb/discussions/1259#discussioncomment-1856422>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AJOKP3UNUH3U2TZF6NXNUMDUSFRRFANCNFSM5KQ2J5QQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Reply by micgresham** [reply]
So, I found the information I wanted on the device by connecting it to a windows machine.  It seems I am not dealing with a pure CDC device, but instead a composite device.  Is there any example code on running host most with a connected composite device?  Attached is the device descriptor information.

thanks,

-michael
[Device Descriptor HPN Serial Port.pdf](https://github.com/hathach/tinyusb/files/7786509/Device.Descriptor.HPN.Serial.Port.pdf)

**Comment by rppicomidi** [comment]
@micgresham If I am reading the descriptor you sent correctly and correctly matched it to the USB class specs, I think the CP2102-9 is a pure CDC PSTN device. See [the class spec](https://www.usb.org/document-library/class-definitions-communication-devices-12). I had a quick look at the spec. I think you get line status on the interrupt endpoint and use bulk IN and OUT endpoints for data.

For the Raspberry Pi Pico's rp2040, bulk endpoints are still being sorted. I submitted a pull request for a MIDI host driver #1219 that supports a bulk IN and bulk OUT endpoint pair like the CP2102-9's descriptor indicates. See @hathach comments on that pull request and also see discussion #1261 for more info. The MIDI host driver lacks the interrupt IN endpoint for line status that the CP2102-9 seems to have.

Hope this helps.

**Reply by micgresham** [reply]
I appreciate.  I really just want to access the virtual comm port (console) on this device.  My working knowledge of USB is limited.
thanks,

-michael

**Comment by IngHK** [comment]
> I want to see what is currently attached to the host controller (rp2040). I will also be testing this with an STMf4xx controller and TinyUSB. Mostly just looking for the basics: Product ID: 0x6001 Vendor ID: 0x0403 (Future Technology Devices International Limited) Version: 6.00 Serial Number: AB0KU7UZ Speed: Up to 12 Mb/s Manufacturer: FTDI I am working with a new CDC interface using the Silicon Labs CP2102 <https://www.silabs.com/documents/public/data-sheets/CP2102-9.pdf> and I am not sure it is coming up properly. It is a USB-C attached serial console built into a piece of equipment. Thanks,
> […](#)
> -michael From: Ha Thach ***@***.***> Reply-To: hathach/tinyusb ***@***.***> Date: Wednesday, December 22, 2021 at 12:52 AM To: hathach/tinyusb ***@***.***> Cc: micgresham ***@***.***>, Author ***@***.***> Subject: Re: [hathach/tinyusb] Is there an function call to show the information about a connected device in host mode? (Discussion #1259) which specific info you want to get ? — Reply to this email directly, view it on GitHub<[#1259 (comment)](https://github.com/hathach/tinyusb/discussions/1259#discussioncomment-1856422)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AJOKP3UNUH3U2TZF6NXNUMDUSFRRFANCNFSM5KQ2J5QQ>. Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. You are receiving this because you authored the thread.Message ID: ***@***.***>

See #2488.
Any kind of feedback, review and test is welcome.
